src/components/Firestore/CollectionFilter.tsx

```diff
@@ -0,0 +1,455 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Button } from '@rmwc/button';
+import { CardActionButton, CardActionButtons, CardActions } from '@rmwc/card';
+import { IconButton } from '@rmwc/icon-button';
+import { CollapsibleList, SimpleListItem } from '@rmwc/list';
+import { Radio } from '@rmwc/radio';
+import { Theme, ThemeProvider } from '@rmwc/theme';
+import { Typography } from '@rmwc/typography';
+import { firestore } from 'firebase';
+import React, { useEffect, useState } from 'react';
+import {
+  Controller,
+  FormContext,
+  useFieldArray,
+  useForm,
+  useFormContext,
+} from 'react-hook-form';
+
+import { grey100 } from '../../colors';
+import { Field, SelectField } from '../../components/common/Field';
+import * as actions from './actions';
+import styles from './CollectionFilter.module.scss';
+import {
+  CollectionFilter as CollectionFilterType,
+  isMultiValueCollectionFilter,
+  isSingleValueCollectionFilter,
+  isSortableCollectionFilter,
+} from './models';
+import { useCollectionFilter, useDispatch } from './store';
+import { isBoolean, isNumber } from './utils';
+
+export const CollectionFilter: React.FC<{
+  className?: string;
+  path: string;
+  onClose?: () => void;
+}> = ({ className, path, onClose }) => {
+  const collectionFilter = useCollectionFilter(path);
+  const dispatch = useDispatch();
+
+  const formMethods = useForm<CollectionFilterType>({
+    mode: 'onChange',
+    defaultValues: collectionFilter,
+  });
+
+  const cf = formMethods.watch({ nest: true });
+
+  const onSubmit = (data: CollectionFilterType) => {
+    dispatch(
+      actions.addCollectionFilter({
+        path,
+        ...data,
+      })
+    );
+    onClose?.();
+  };
+
+  return (
+    <CollectionFilterTheme>
+      <FormContext {...(formMethods as any)}>
+        <form
+          onSubmit={formMethods.handleSubmit(onSubmit)}
+          className={className}
+        >
+          {/* Field entry */}
+          <FilterItem title="Filter by field" preview={cf.field} defaultOpen>
```

Comment:
I love your layers of abstraction here. The fact that the button UI at the bottom is only slightly less lines compared to the content here for the main part of the form ❤️ 

Makes this easy to follow.

Reply:
Thanks Tim!